src/main/kotlin/tanvd/grazi/ide/ui/GraziSettingsPanel.kt

```diff
@@ -16,36 +19,66 @@ import org.picocontainer.Disposable
 import tanvd.grazi.GraziConfig
 import tanvd.grazi.ide.ui.rules.GraziRulesTree
 import tanvd.grazi.language.Lang
-import java.awt.Desktop
 import javax.swing.JComponent
 import javax.swing.JEditorPane
-import javax.swing.event.HyperlinkEvent
+import javax.swing.JLabel
 
 class GraziSettingsPanel : ConfigurableUi<GraziConfig>, Disposable {
     private val cbEnableGraziSpellcheck = JBCheckBox(msg("grazi.ui.settings.enable.text"))
     private val cmbNativeLanguage = ComboBox<Lang>(Lang.sortedValues.toTypedArray())
-    private val adpEnabledLanguages by lazy { GraziAddDeleteListPanel() }
 
-    private val descriptionPane = JEditorPane().apply {
+    private val ruleLink = LinkLabel<Any?>(msg("grazi.ui.settings.rules.rule.description"), null)
+    private val linkPanel = panel(HorizontalLayout(0)) {
+        border = padding(JBUI.insetsBottom(10))
+        isVisible = false
+
+        add(ruleLink)
+        add(JLabel(AllIcons.Ide.External_link_arrow))
+    }
+
+    private val smallInfoPane = JEditorPane().apply {
         editorKit = UIUtil.getHTMLEditorKit()
         isEditable = false
         isOpaque = true
         border = null
         background = null
+    }
 
-        addHyperlinkListener { event ->
-            if (event?.eventType == HyperlinkEvent.EventType.ACTIVATED) {
-                Desktop.getDesktop()?.browse(event?.url?.toURI())
-            }
-        }
+    private val descriptionPane = JEditorPane().apply {
+        editorKit = UIUtil.getHTMLEditorKit()
+        isEditable = false
+        isOpaque = true
+        border = null
+        background = null
     }
 
     private val rulesTree by lazy {
         GraziRulesTree {
+            smallInfoPane.text = getSmallInfoPaneContent(it)
+
+            linkPanel.isVisible = getLinkLabelListener(it)?.let { listener ->
+                ruleLink.setListener(listener, null)
+                true
+            } ?: false
+
             descriptionPane.text = getDescriptionPaneContent(it)
         }
     }
 
+    private val adpEnabledLanguages by lazy {
+        GraziAddDeleteListPanel(object : GraziLanguagePanelUpdateListener {
+            override fun onLanguageAdded(lang: Lang) {
+                rulesTree.langs.add(lang)
+                rulesTree.update()
+            }
+
+            override fun onLanguageRemoved(lang: Lang) {
+                rulesTree.langs.remove(lang)
+                rulesTree.update()
```

Comment:
Should not rulesTree update itself on add and remove of lang by itself? If you will encapsulate change of set of langs into methods you may move there `update` too.

Reply:
Done